Sources/Kafka/RDKafka/RDKafkaClient.swift

```diff
@@ -124,47 +124,193 @@ final class RDKafkaClient: Sendable {
+
+            arguments[index].u.header.name = cHeader.key
+            arguments[index].u.header.val = cHeader.value?.baseAddress
+            arguments[index].u.header.size = cHeader.value?.count ?? 0
+
+            index += 1
+        }
+
+        assert(arguments.count == size)
+
+        return rd_kafka_produceva(
+            self.kafkaHandle,
+            arguments,
+            arguments.count
+        )
+    }
+
+    /// Scoped accessor that enables safe access to a ``KafkaProducerMessage``'s key and value raw buffers.
+    /// - Warning: Do not escape the pointer from the closure for later use.
+    /// - Parameter body: The closure will use the pointer.
+    @discardableResult
+    private static func withMessageKeyAndValueBuffer<T, Key, Value>(
+        for message: KafkaProducerMessage<Key, Value>,
+        _ body: (UnsafeRawBufferPointer?, UnsafeRawBufferPointer) throws -> T // (keyBuffer, valueBuffer)
+    ) rethrows -> T {
+        return try message.value.withUnsafeBytes { valueBuffer in
+            if let key = message.key {
+                return try key.withUnsafeBytes { keyBuffer in
+                    return try body(keyBuffer, valueBuffer)
+                }
+            } else {
+                return try body(nil, valueBuffer)
+            }
+        }
+    }
+
+    /// Scoped accessor that enables safe access the underlying memory of an array of ``KafkaHeader``s.
+    /// - Warning: Do not escape the pointer from the closure for later use.
+    /// - Parameter body: The closure will use the pointer.
+    @discardableResult
+    private static func withKafkaCHeaders<T>(
+        for headers: [KafkaHeader],
+        _ body: ([(key: UnsafePointer<CChar>, value: UnsafeRawBufferPointer?)]) throws -> T
+    ) rethrows -> T {
+        var headersMemory: [(key: UnsafePointer<CChar>, value: UnsafeRawBufferPointer?)] = []
+        var headers: [KafkaHeader] = headers.reversed()
+        return try self._withKafkaCHeadersRecursive(kafkaHeaders: &headers, cHeaders: &headersMemory, body)
+    }
+
+    /// Recursive helper function that enables safe access the underlying memory of an array of ``KafkaHeader``s.
+    /// Reads through all `kafkaHeaders` and stores their corresponding pointers in `cHeaders`.
+    private static func _withKafkaCHeadersRecursive<T>(
+        kafkaHeaders: inout [KafkaHeader],
+        cHeaders: inout [(key: UnsafePointer<CChar>, value: UnsafeRawBufferPointer?)],
+        _ body: ([(key: UnsafePointer<CChar>, value: UnsafeRawBufferPointer?)]) throws -> T
+    ) rethrows -> T {
+        if kafkaHeaders.isEmpty {
+            // Base case: we have read all kafkaHeaders and now invoke the accessor closure
+            // that can safely access the pointers in cHeaders
+            return try body(cHeaders)
+        } else {
+            guard let kafkaHeader = kafkaHeaders.popLast() else {
+                fatalError("kafkaHeaders should not be nil")
```

Comment:
Can we swap this around so that we only guard. Then we don't need the `fatalError` at all